packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -514,6 +503,50 @@ function getBackCompatRuntimeOptions(runtimeOptions?: IContainerRuntimeOptions):
     return { summaryOptions, gcOptions };
 }
 
+/**
+ * Produce a ContainerRuntime with AgentScheduler, compatible with ContainerRuntime.load() from versions prior to 0.38.
+ * @deprecated This is provided as a migration tool only.  If you require back compat with documents produced prior to
+ * 0.38, you'll need to add AgentScheduler to your ContainerRuntime registry.  If you want AgentScheduler
+ * functionality, you should instantiate it in your container code as your scenario demands.
+ */
+export const makeLegacyContainerRuntime = async (
+    context: IContainerContext,
+    registryEntries: NamedFluidDataStoreRegistryEntries,
+    requestHandler?: (request: IRequest, runtime: IContainerRuntime) => Promise<IResponse>,
+    runtimeOptions?: IContainerRuntimeOptions,
+    containerScope?: IFluidObject,
+) => {
+    const augmentedRegistry = [
+        ...registryEntries,
+        AgentSchedulerFactory.registryEntry,
+    ];
+
+    const runtime = await ContainerRuntime.load(
+        context,
+        augmentedRegistry,
+        requestHandler,
+        runtimeOptions,
+        containerScope,
+    );
+
+    // Create all internal data stores if not already existing on storage or loaded a detached
+    // container from snapshot(ex. draft mode).
+    if (!context.existing) {
+        await runtime.createRootDataStore(AgentSchedulerFactory.type, agentSchedulerId);
+    }
+
+    if (context.clientDetails.capabilities.interactive) {
+        const scheduler = await requestFluidObject<IAgentScheduler>(
+            await runtime.getRootDataStore(agentSchedulerId, true),
+            "",
+        );
+        const leadershipManager = new LeadershipManager(scheduler);
```

Comment:
where and how leadershipManager is exposed to caller?
Should we instead provide an accessor to scheduler (in some form or shape), such that individual callers can get tasks with unique keys?

Reply:
In this approach it's safe to instantiate multiple `LeadershipManager`s for a single `AgentScheduler`, so anywhere someone wants to use it can instantiate their own.  There are many options we could take here -- this one optimizes for being as close in behavior to current behavior while still removing public API surface from the `ContainerRuntime`, but we could choose other tradeoffs.

Reply:
That make sense, but it requires knowledge that agent scheduler is available, and how to get to it. I.e. some helper method would be useful here. I think we should explore if we can make it a class (not a function makeLegacyContainerRuntime), such that accessor to it would be on such class. Or return agent scheduler from makeLegacyContainerRuntime (in addition to runtime).

I like approach of mixins, but that requires removing of static ContainerRuntime.load() as a creation path

Reply:
Bohemia already does `containerRuntime.request({ url: '/_scheduler' })` in several places which seems OK to me for when we're talking about the legacy one (new users of `AgentScheduler` can try other strategies, e.g. storing a handle somewhere).

Reply:
The other approach I considered (in "4.1" from our conversation) was to augment the `ContainerRuntime` itself before handing it back.  I think it's easier to make the shift to unique keys with this approach ("4.2") but that's another option.

Reply:
Another option would be to pass the `TaskSubscription` (or the `AgentScheduler` itself) through the container scope.  Or to register a specific request handler on the container to request the leadership `TaskSubscription` (e.g. `containerRuntime.request({ url: '/leadershipTaskSubscription' });`)